### ISSUE TITLE
NoCredentialProviders: no valid providers in chain. Deprecated. 

### ISSUE THREAD

**Issue by sbenthall** [apex/apex#426]
Ubuntu 14.04, Apex version 0.9.0

Running `apex metrics` gets me the following error, looping in the terminal.

```
NoCredentialProviders: no valid providers in chain. Deprecated. 
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

I am not a Go developer and so am not sure how to interpret or fix this kind of error. A more informative message would be helpful.

I have my AWS credential in `~/.aws/config`

**Comment by tj** [comment]
hmm I haven't seen that one yet, _sounds_ like a dev-error as if we misconfigured something, but it works fine for me. I'll look into it, could just be strange wording on their part.

note that you'll need creds in ~/.aws/credentials, only the region is stored in ~/.aws/config (not sure why AWS likes to have two but that seems to be the convention)

**Comment by sbenthall** [comment]
thanks moving the creds to ~/.aws/credentials fixed the issue for me.

**Comment by tj** [comment]
cool good to hear, I'll make sure to get better error reporting for this in the next release

**Comment by louisremi** [comment]
I also got this error because of a syntax error in my .envrc file (a space just after `AWS_SECRET_KEY=`)

**Comment by rowanu** [comment]
Another data point for this error: I'm using an assumed role, so there's (intentionally) no credentials in `~/.aws/credentials` file - it all goes [in the `~/.aws/config` file](http://docs.aws.amazon.com/cli/latest/userguide/cli-roles.html).

Is this Apex error just caused by an SDK error? I know the SDK support can be a bit spotty...

**Comment by DavidTPate** [comment]
Yeah, this is just an SDK error where upon initialization it attempts to determine your credentials (using EC2 role, assumed role, credentials, environment variables, etc.) and when it can't determine them then it throws the error.

**Comment by dpurrington** [comment]
This is happening on Ubuntu. I installed the AWS CLI using apt-get. After running "aws configure" I was left with a single ~/.aws/config file. Separating into two files (config and credentials) is a workaround, but that breaks the CLI. You should probably consider these files to be an implementation detail of the AWS SDK. I think the SDK should do all the searching for creds for you, so it shouldn't be necessary to have this dependency.

**Comment by grauwoelfchen** [comment]
Hi, I also got this error in a situation that I had both files `~/.aws/config` for region and `~/.aws/credentials` for credentials, but didn't have `AWS_SDK_LOAD_CONFIG` environment variable as truthy, yet.

All of you might already know well about this... so, I would like to just report my situation.
If you have both of these files, you need to also set `AWS_SDK_LOAD_CONFIG` as truthy  such as described at [here](https://github.com/aws/aws-sdk-go#aws-shared-config-file-awsconfig).

I also think that it would be better aws-sdk-go reports this problem effectively ;)

**Comment by davidcpell** [comment]
Just landed here and @grauwoelfchen's point is what got me moving again. Thanks!

**Comment by odg0318** [comment]
@grauwoelfchen Thanks a lot.

**Comment by xuru** [comment]
@grauwoelfchen's suggestion worked for me as well.  This should probably be documented somewhere, especially for people coming over from using something like the aws cli.

**Comment by VolkerBohne** [comment]
@grauwoelfchen : thx a lot. helped me to get rid of that error when beginnnig with apex and cross account access. cool.

**Comment by Pczek** [comment]
I still encounter this issue. I followed the instruction on http://apex.run about the credentials and created a credentials and a config file. No additional whitespaces at the end as mentioned. I even set ```AWS_SDK_LOAD_CONFIG``` to 1 as described in the readme in the aws-sdk-go project.
Might there be something I need to install additionally?

**Comment by cynipe** [comment]
me too.
0.16.0 does not work as expected.
However, 0.15.0 works fine.

**Comment by amantur** [comment]
Folks, I got bitten by  `* provider.aws: NoCredentialProviders: no valid providers in chain.` bug while running a terraform script. In the end it  turned out to be a typo in configured profile name and what I had set in the configuration file.

**Comment by azarboon** [comment]
@Pczek thank you. only your solution (adding env) worked for me.

**Comment by DrHashi** [comment]
I was only able to get it working using environment variables as of aws.provider 1.26

**Comment by michele-amati** [comment]
> Folks, I got bitten by `* provider.aws: NoCredentialProviders: no valid providers in chain.` bug while running a terraform script. In the end it turned out to be a typo in configured profile name and what I had set in the configuration file.

1. look for solutions on the internet
2. the internet tells me i typed it wrong
3. check the names for the third time
4. ...i typed it wrong
5. thank you internet

**Comment by stephanebruckert** [comment]
Just the wrong `-p profile` AWS profile for me

**Comment by nunofernandes** [comment]
Also hit by this (vault version 1.1.0 SHA 36aa8c8dd1936e10ebd7a4c1d412ae0e6f7900bd) . When I try to run it without ECS task role, I get the expected:

```
Error parsing Seal configuration: error fetching AWS KMS sealkey information: AccessDeniedException: User: arn:aws:sts::XXXX:assumed-role/svc-spot-ecs-instance-role/i-0c27bb6af7118f355 is not authorized to perform: kms:DescribeKey on resource: arn:aws:kms:eu-west-1:XXX:key/123456789
```

That is to be expected as the instance role doesn't have any permissions on that KMS key. When running inside the ECS task (that has the permission for KMS) with the AWS_CONTAINER_CREDENTIALS_RELATIVE_URI environment variable defined, I the the error:

"Error parsing Seal configuration: error fetching AWS KMS sealkey information: NoCredentialProviders: no valid providers in chain. Deprecated"

Thanks for any pointers on debugging this.

**Comment by kolly83** [comment]
Hello

I have my credentials set in the ~/.aws/credentials but when I run the packer to build the AMI use the json file I've still got the below error:

env AWS_PROFILE=packer-demo packer build -only amazon-ebs packer-demo.json

amazon-ebs output will be in this color.

==> amazon-ebs: Prevalidating AMI Name...
==> amazon-ebs: Error querying AMI: NoCredentialProviders: no valid providers in chain. Deprecated. 
==> amazon-ebs: 	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
Build 'amazon-ebs' errored: Error querying AMI: NoCredentialProviders: no valid providers in chain. Deprecated. 
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: Error querying AMI: NoCredentialProviders: no valid providers in chain. Deprecated. 
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors

==> Builds finished but no artifacts were created.


Any has got any idea?

Thank you.
